File: frontend/src/Forecast/ForecastApi.ts
import {Forecast} from './ForecastState';
import * as schemawax from 'schemawax';
import {Http, http} from '../Http/Http';
import {Result} from '../Http/Result';
import {ForecastRequest} from '../Teams/ForecastRequestState';

const teamDecoder: schemawax.Decoder<{ name: string }> = schemawax.object({required: {name: schemawax.string}});

const forecastDecoder: schemawax.Decoder<Forecast> =
    schemawax.object({
        required: {
            fixture: schemawax.object({
                required: {
                    home_team: teamDecoder,
                    away_team: teamDecoder,
                    league: schemawax.string,
                }
            }),
            outcome: schemawax.literalUnion('home', 'away', 'draw'),
            model_name: schemawax.string
        },
        optional: {
            confidence: schemawax.nullable(schemawax.number)
        }
    }).andThen((json): Forecast => ({
        fixture: {
            home: {name: json.fixture.home_team.name, leagues: [json.fixture.league]},
            away: {name: json.fixture.away_team.name, leagues: [json.fixture.league]},
            league: json.fixture.league,
        },
        model_name: json.model_name,
        outcome: json.outcome,
        confidence: json.confidence || undefined,
    }));

const fetchFor = (request: ForecastRequest): Promise<Result<Forecast, Http.Error>> => {
    const params = new URLSearchParams({
        home_name: request.home.name,
        away_name: request.away.name,
        league: request.league,
        model_name: request.model.name,
    });

    if (request.matchStatus.type == 'in progress') {
        params.append('home_goals', request.matchStatus.homeGoals.toString());
        params.append('away_goals', request.matchStatus.awayGoals.toString());
        params.append('minutes_elapsed', request.matchStatus.minutesElapsed.toString());

        return http.sendRequest(`/api/forecast-in-progress?${params}`, forecastDecoder);
    } else {
        return http.sendRequest(`/api/forecast?${params}`, forecastDecoder);
    }
};

export const forecastApi = {
    fetchFor,
};

File: backend/matchpredictor/forecast/forecast_api.py
from flask import Blueprint, jsonify, request, Response

from matchpredictor.forecast.forecaster import Forecaster
from matchpredictor.matchresults.result import Team, Fixture, Scenario


def forecast_api(forecaster: Forecaster) -> Blueprint:
    api = Blueprint("forecast_api", __name__)

    @api.route("/forecast", methods=["GET"])
    def forecast() -> Response:
        home_name = request.args['home_name']
        away_name = request.args['away_name']
        league = request.args['league']
        model_name = request.args['model_name']

        result = forecaster.forecast(
            Fixture(
                home_team=Team(name=home_name),
                away_team=Team(name=away_name),
                league=league,
            ),
            model_name=model_name,
        )

        if result is None:
            return Response("Cannot forecast fixture", 400)

        return jsonify(result)

    @api.route("/forecast-in-progress", methods=["GET"])
    def forecast_in_progress() -> Response:
        home_name = request.args['home_name']
        away_name = request.args['away_name']
        league = request.args['league']
        model_name = request.args['model_name']
        minutes_elapsed = request.args.get('minutes_elapsed', default=0, type=int)
        home_goals = request.args.get('home_goals', default=0, type=int)
        away_goals = request.args.get('away_goals', default=0, type=int)

        result = forecaster.forecast_in_progress(
            Fixture(
                home_team=Team(name=home_name),
                away_team=Team(name=away_name),
                league=league,
            ),
            Scenario(
                minutes_elapsed=minutes_elapsed,
                home_goals=home_goals,
                away_goals=away_goals,
            ),
            model_name=model_name,
        )

        if result is None:
            return Response("Cannot forecast fixture", 400)

        return jsonify(result)

    return api
